### ISSUE TITLE
Cannot have the same node for two different inputs to the same WorkChain

### ISSUE THREAD

**Issue by greschd** [aiidateam/aiida_core#641]
When two different inputs to the same ``WorkChain`` are given the same input, it fails with ``UniquenessError: A link from node ... and the current node ... already exists!``. In my opinion this is a bug, since having two inputs which are equal certainly doesn't break the DAG. For most cases we can just get around this issue by duplicating the input node, but if the input node is something like a code that is a bit ridiculous.

What's the reason for raising this exception? IMO, one could just skip the link creation if a link of the same type (and direction) already exists between the two nodes, without raising an error.

Branch with failing test: https://github.com/greschd/aiida_core/tree/double_link

**Comment by sphuber** [comment]
I think duplicating the input node is certainly the wrong approach. I take it this problem arises in your wrapper workchains, where the same node is used by multiple subworkchains and is therefore passed under different keys to the main workchain? As you say this is a perfectly valid situation from the workchains perspective, but this duplication should not be visible on the DAG level. That is to say, it should just be a single input node. It should not matter that that same node is used multiple times within the workchain. I see that the link also gets a label, so we could say, that we change the uniqueness condition to be uuid+label, the label being the input key in the case of the workchain, but I would even say that removing that constraint would even be fine. Adding the same node with the same input link label, multiple times does not break anything if we just skip it right? Maybe we should simply change the exception to a warning. The only case I can see this be a problem is where a user made a mistake and accidentally attached the same node twice.

**Comment by greschd** [comment]
Exactly, I was passing the same ``Code`` under two different keys corresponding to different parts of the workflow.

Should the linktype also be part of the uniqueness condition? That is, does it make sense to have links of different types between two nodes?

I propose the following behaviour:
* same uuid, label and linktype: skip (do nothing)
* otherwise: add new link 

Or should there be some conditions where we still keep the error?

**Comment by sphuber** [comment]
I just thought of the link type discussion, which will directly influence the correct solution here. I have been working on a design for the provenance and link types and in that design the problem would be slightly different. In that conception, workchain nodes will be different node types from calculation nodes and their link types will be different. If we were to go with that setup, it would not make sense to have multiple links from the same input node to the same calculation, so the exception raised could stay. But for work nodes, having multiple input links would be fine. Since we would have well defined node and link types, we can simply add that to the check. It would look something like:

* `Data -> Calc:  Uniqueness(Data.uuid)`
* `Data -> Work: Uniqueness(Data.uuid, Link.label)`

The same would go for output data nodes. So now I am not sure how to fix this problem in the current branch. With the current definition of node and link types, we may have to keep the uniqueness constraint.

**Comment by greschd** [comment]
Is there any particular reason why calculations shouldn't be able to have the same node as input twice?

**Comment by greschd** [comment]
Right now there are actually **two** checks: 
* Check 1: If the same label is already present
* Check 2: If a link from the same UUID exists already

I think we can just drop Check 2. Having two links with the same label (even with different linktype or UUID) doesn't really work because then we overwrite the key in the link-cache dictionary.

**Comment by greschd** [comment]
Another question is whether the same change should also be done in ``_replace_link_from``.

**Comment by greschd** [comment]
The only time when ``_replace_link_from`` is used is in the ``use_*`` methods. As far as I can tell, it's so that you could call the same method twice (which would then just replace the input link). So I think it should have the same behaviour, also allowing multiple links between two nodes.

**Comment by sphuber** [comment]
I think the first check is totally wrong regardless. Why shouldn't you be able to have an input and output link from a certain link that both have the label 'output_parameters'. It might be weird for the input node, but should be perfectly legal. What we really should be checking for I think is the uniqueness of the `(src.uuid, link_label)` tuple.

Martin and I actually fixed this some time ago by improving the structure of the `input_links_cache`, but this was in our massive workchains branch and I forgot to backport it. I just discussed with @muhrin and we think we should just remove both checks and only check for the uuid/label tuple. The only problem is that some other code may sadly bank on this old constraint but maybe does not have explicit tests for. So if we were to change this, tests might pass, while some stuff may still be broken.

Our proposed solution is then to

* Loosen uniqueness requirement to `(src.uuid, label)`
* Write unit tests that inscribe and test this behavior

**Comment by greschd** [comment]
Yeah, the only reason I saw for not allowing the same label was indeed that it's not just a trivial change in the implementation -- but if you've anyway done this already, then it makes sense to go that way.

Maybe we can look through the code if there's a try-catch for ``UniquenessError`` anywhere to see if someone relies on this behaviour? End users shouldn't really care about this, I wouldn't expect them to create the links manually.

**Comment by sphuber** [comment]
No but I am worried that maybe for example something like the `QueryBuilder` in some parts assumes that if it fetches the links between two nodes, it only ever will return at most a single link. That would of course be broken and it might that tests do not capture this.

You want me to try and backport our changes and change the check to the tuple, or do you want to keep your PR?

**Comment by greschd** [comment]
Yeah that could indeed be a problem -- maybe we should talk to @lekah about this?

If you think your backport solves it better, then I'm fine with closing my PR -- I can still use it to make my code work in the meantime, anyway :slightly_smiling_face:

**Comment by sphuber** [comment]
I went back to the fix I mentioned in our old workchain branch and started backporting it, but half way through I realized that we missed several places that accessed the `_input_links_cache` yet as far as I know, the tests ran just fine :cold_sweat:

So now I am not comfortable backporting that fix even though some parts of it are useful. I thought about it some more and in the current definition of links and link types, I think we have the following set of link uniqueness constraints. Following table considers the action of wanting to add a link from the current node to the other node with label and type. In the case where each field already matches (+) or is different (-) from existing links, is the new link allowed?

node | label | type | Allowed? | Remark
----------|---------|--------|---------|---------
\+   |    +    |    +   |   :x:  | Perfect duplicate does not work of course
\+   |    +    |    -   |   :white_check_mark: | Not 100% sure but I think this should be fine. Only link type combinations that would be possible are input/return, input/call and call/return. All of those would be fine with identical node and label
\+   |    -     |    +   |   :white_check_mark: | Two input links from same node with different label, this is what was causing your problem
\-   |    +    |    +   |   :x: | Two input links from different nodes but same label would cause problems

What this boils down to is that we have to do two simple checks I think in `add_link_from`.

    # If link exists with all three fields identical, the link already exists, so job is done
    if link(src, label, link_type) in self._input_links_cache:
        return

    # This is not allowed
    if link(label, type) in self._input_links_cache:
        raise UniquenessError()

What do you reckon?

**Comment by greschd** [comment]
Makes sense to me. So the two conditions you are checking are the two with are "not allowed", with the first one just returning, and the second one raising the error, correct? And the two cases which are allowed just go on to create the link as usual?

With the current implementation the only trouble is that (I think) we can't have two links of different type with the same label (again, because it's the key of the ``_input_links_cache``). Should we implement it such that those are also not allowed as a workaround for now, or do you want to have another go at backporting? I think it should be possible to find all the places just with a global search for ``_input_links_cache``. It's also possible that the places you missed are just dead code...

**Comment by sphuber** [comment]
The problem was that the fix required they keys in `_input_links_cache` to be at least a tuple of `(src.uuid, label)`. This is also exactly what we implemented, but then I found methods like `remove_link` that only passed the `label` as an argument. At that point I stopped, but it might be that those methods are indeed dead code, or it would be simple enough to adapt them to pass all required information. I haven't deleted the backport yet, so I will have another look tomorrow to see how much has to be adapted. I just didn't want to invest too much time in something that has to be changed altogether soon anyway, while still running the risk of breaking develop.
